Regenerate starter app using Vite

dargaCode opened the pull request:
closes #4 

I used a [community template](https://github.com/fabien-ml/react-ts-vite-template) to generate the boilerplate. It's a lot of stuff, but my thinking was it'd be good to at least demo an example of all these tools, even if the team doesn't end up using them much or at all.

I kept the fetching `/utils` and the original CtD template logic, and added some more little stuff like
- Vite environment variable example
- `@` aliases for import paths
- A little extra error handling to nudge us toward the fullstack dev workflow

<img width="250" alt="Screen Shot 2023-10-24 at 5 12 30 PM" src="https://github.com/Code-the-Dream-School/ee-prac-team2-front/assets/12771651/993c979e-0d29-4733-b68f-5e3ff28a3931">

# What's the stack?
#### Build
- [Vite](https://vitejs.dev/)
#### Dev
 - [ReactJS](https://reactjs.org/)
- [Typescript](https://www.typescriptlang.org/)
#### Lint
- [ESLint](https://eslint.org/)
- [Prettier](https://prettier.io/)
#### Test
- [Jest](https://jestjs.io/)
- [Testing Library](https://testing-library.com/)
- [Cypress](https://www.cypress.io/)
#### Ops
- [Husky](https://typicode.github.io/husky/)
- [CommitLint](https://commitlint.js.org/#/)
- [Commitizen](https://commitizen-tools.github.io/commitizen/)
   - This is the only one I haven't used, might be overkill XD

# What's left `TODO`?
(can convert these to issues)
#### Stack
- Add [Storybook](https://storybook.js.org/)
#### Dev
- Review/replace ESlint rules?
- Rename/simplify npm scripts?
#### Code Review
- PR template
- Pre-merge checks via Actions

dargaCode committed:
build: rebuild app using Vite template

combined with some of the existing logic from the CtD version

new template includes lots of other utilities eg typescript, cypress,
husky, jest, TODO need to add storybook

dargaCode committed:
chore(build): add better attribution

dargaCode commented on .eslintrc.json:
```diff
@@ -0,0 +1,113 @@
+    "simple-import-sort" // Plugin for sorting imports in file.
+  ],
+  "rules": {
+    "prettier/prettier": "off",
```
didn't touch ESlint at all yet except to silence warnings from Prettier. They'll be fixed automatically via the editor or a script, but they won't clutter up the linter when they're "wrong"

dargaCode commented on .prettierrc.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "arrowParens": "always",
+  "htmlWhitespaceSensitivity": "ignore"
```
I did modify both these Prettier rules:
- https://prettier.io/docs/en/options.html#arrow-function-parentheses
   - (always place optional perens `(like) => "this"`
- https://prettier.io/docs/en/options.html#html-whitespace-sensitivity
   - I believe this stops it from adding extra `{" "}` whenever we reformat `display:inline` elements

dargaCode committed:
fix(assets): add back readme images

dargaCode commented on jest.setup.ts:
```diff
@@ -2,4 +2,8 @@
-import '@testing-library/jest-dom';
+import "@testing-library/jest-dom";
+
+beforeEach(() => {
+  expect.hasAssertions();
+});
```
Make sure a test which doesn't remember to actually test anything won't incorrectly show as passing 🙃

dargaCode committed:
fix(package): add back in `private` entry

dargaCode committed:
fix(build): fix favicon path

dargaCode commented on src/App.tsx:
```diff
@@ -0,0 +1,20 @@
+import Logo from "@assets/logo.png";
+import DataContainer from "@components/DataContainer/DataContainer";
+import HelloWorld from "@components/HelloWorld/HelloWorld";
```
Using the Typescript/Vite path aliases to prevent lots of `../../../`

dargaCode commented on src/App.tsx:
```diff
@@ -0,0 +1,20 @@
+      <img className={styles.logo} alt="React logo" width="400px" src={Logo} />
+      <DataContainer />
+      <div>
+        {`VITE_SOME_ENVIRONMENT_VARIABLE from .env.local: "${
```
Vite has a different way of accessing environment variables than Create React App (which uses Webpack behind the scenes)

dargaCode commented on src/components/DataContainer/DataContainer.tsx:
```diff
@@ -0,0 +1,26 @@
+export default function DataContainer() {
+  const [message, setMessage] = useState("");
+
+  useEffect(() => {
+    (async () => {
+      const response = await getAllData(URL);
+
+      setMessage(response.data);
+    })();
```
TODO replace this `async` IIFE with a `Promise.then()` for readability?

dargaCode commented on src/components/DataContainer/DataContainer.tsx:
```diff
@@ -0,0 +1,26 @@
+
+const URL = "http://localhost:8000/api/v1/";
+
+export default function DataContainer() {
```
When merging the two templates' logic together, I put the CtD stuff in this component

dargaCode commented on src/main.tsx:
Looks like Vite build defaults to `main.tsx` instead of `index.tsx`, but we can easily change back if we want!

dargaCode commented on src/reportWebVitals.js:
I've not used this before but we can definitely add back if needed

dargaCode commented on src/utils/index.ts:
`/util` -> `utils` bc I couldn't help myself, but I can revert for consistency with the other projects if we like 😅

dargaCode commented on src/utils/index.ts:
```diff
@@ -0,0 +1,28 @@
+  } catch (error) {
+    console.log(error, `error - getAllData in ${url} route`);
+
+    if (axios.isAxiosError(error)) {
+      return {
+        data: "API request failed! Did you remember to `npm start` the backend app?",
+      };
+    }
```
Converted these functions to Typescript as well as hinting to start the backend app if there was an error

angelkbrown replied:
Heads up that there's a small typo here -- `npm run dev` starts the backend app, rather than `npm start`.

dargaCode replied:
fixed, thank you!

dargaCode commented on tsconfig.eslint.json:
```diff
@@ -0,0 +1,18 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    // ensure that nobody can accidentally use this config for a build
```
i guess we can put comments into json without breaking it now? this template seems to do that a lot

dargaCode commented on vite.config.ts:
```diff
@@ -0,0 +1,26 @@
+export default defineConfig({
+  plugins: [legacy(), react()],
+  resolve: {
+    alias: {
+      "@assets": path.resolve(__dirname, "./src/assets"),
+      "@components": path.resolve(__dirname, "./src/components"),
+      "@utils": path.resolve(__dirname, "./src/utils"),
+    },
```
adding the `@` path aliases to Vite.

dargaCode commented on tsconfig.json:
```diff
@@ -0,0 +1,28 @@
+    "sourceMap": true,
+    "types": ["vite/client", "node", "jest", "@testing-library/jest-dom"],
+    "baseUrl": ".",
+    "paths": {
+      "@assets/*": ["src/assets/*"],
+      "@components/*": ["src/components/*"],
+      "@utils": ["src/utils"]
+    }
```
adding the `@` path aliases to Typescript. 

(`"@utils"` instead of `"@utils/*"` because `/utils` uses an `index.js`. TODO we can consider also using `index.jsx` for `/components`)

angelkbrown reviewed (dismissed):
Wow! This all looks great to me, Darga! I ran the app locally successfully ran the cypress spec as well (I think -- this was my first experience with Cypress!)

I'll comment on your open question items in the PR description as well. Thanks so much for doing this! 🎉 🤩

angelkbrown commented:
Re: "What's left todo"

> Stack
> 
>     Add [Storybook](https://storybook.js.org/)
>     Probably don't add [Next.js](https://nextjs.org/)?

I'm not familiar with Storybook -- is it sort of a generator for fancy components? My instinct is it would be good for folks to get more experience building components from scratch before more abstraction is added, but we can definitely talk through this as a group if you feel this is important for them to learn.

My understanding is that Next.js is a fullstack framework. Since we have a separate backend, I agree it wouldn't make sense to add Next.js to our frontend.

> Dev
> 
>     Review/replace ESlint rules?
>     Rename/simplify npm scripts?

👍 

> Branches
> 
>     main
>         Change to develop?
>         Disallow push
>         Require PR to merge
>     staging?
>     production?

It looks like the suggested workflow [here](https://learn.codethedream.org/resources/git-process-for-code-the-dream-practicums-b/) is that there's a `dev` branch (separate from `main`) that feature branches are merged to and then periodically mentors merge `dev` into `main`. Since we'll probably just have our local environments and the production environment, I don't think we'll have a need for `staging` and `production` branches (since production will match whatever's in `main` and local environments can use `dev` or feature branches.) Does that all sound right to you? Happy to discuss more or with the team. :)

> Code Review
> 
>     PR template
>     Minimum approvals to merge
>         Start with One mentor?
>         Evolve to One mentor + Two teammates?
>             Maybe too strict, but emphasizes code review?
>     Pre-merge checks via Actions
>     Automatic Reviews via code owners?
>         Or would we need to ask CtD to create a team?

All this sounds like a good place to start! We can iron it out as we go. 👍 

> Build/Release
> 
>     Process/tool?
>         This has never been something I've been very involved in

I think this will depend on how we deploy. I looked a bit at AWS elastic beanstalk deploys today. I suspect it may be a good bit more involved than deploying to [Render](https://learn.codethedream.org/resources/git-process-for-code-the-dream-practicums-b/) as suggested. We can talk more about this as a team, but I'm wondering if it may make sense to deploy to Render to start, and have AWS deployment as a stretch goal. If we deploy to Render, deployments are as easy as pushing to `main`, and it looks like doing things like [making successful test runs a deploy dependency](https://render.com/docs/deploy-hooks#example-workflow) is straightforward as well.

dargaCode committed:
chore(package): add package description to match backend repo

dargaCode committed:
Merge remote-tracking branch 'origin/main' into rebuild-via-vite

dargaCode committed:
fix(readability): remove IIFE

dargaCode committed:
fix: backend start command

dargaCode commented:
> Re: "What's left todo"

This all sounds great, thank you! I'll send you some more info on slack about Storybook

dargaCode commented:
Re-requesting review after adding `CODEOWNERS` and changing base branch

dargaCode committed:
remove commitlint and commitizen packages

These are too strict and not actually useful in the context of this type of project

dargaCode commented:
Bypassing the extra review so that I can have this checked in and ready for people to pull at the team meeting!
